### PR TITLE
pfft: init at 1.0.8-alpha; pnfft: init at 1.0.7-alpha

### PR DESCRIPTION
**Pull request by hmenke**
## Description of changes

This contributes the two parallel FFT packages developed by TU Chemnitz.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/pn/pnfft/package.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    export FCFLAGS="-I${lib.getDev fftw'}/include -I${lib.getDev pfft'}/include"
+    bash bootstrap.sh
+  '';
+
+  configureFlags = [
+    "--enable-threads"
+    "--enable-portable-binary"
+  ] ++ lib.optional (precision != "double") "--enable-${precision}";
+
+  buildInputs = [ gsl ] ++ lib.optionals stdenv.cc.isClang [ llvmPackages.openmp ];
```
Comment (markuskowa) [inline pkgs/by-name/pn/pnfft/package.nix:47]:
```suggestion
  buildInputs = [ gsl ] ++ lib.optional stdenv.cc.isClang llvmPackages.openmp;
```

File: pkgs/by-name/pn/pnfft/package.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    autoconf
+    automake
+    libtool
+  ];
+
+  preConfigure = ''
+    export FCFLAGS="-I${lib.getDev fftw'}/include -I${lib.getDev pfft'}/include"
```
Comment (markuskowa) [inline pkgs/by-name/pn/pnfft/package.nix:38]: the include paths should be picked up automatically, when you put `fftw'` and `pfft'` in `buildInputs`?

File: pkgs/by-name/pn/pnfft/package.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    url = "https://www-user.tu-chemnitz.de/~potts/workgroup/pippig/software/pnfft-${finalAttrs.version}.tar.gz";
+    hash = "sha256-/aVY/1fuMRl1Q2O7bmc5M4aA0taGD+fcQgCdhVYr1no=";
+  };
+
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    autoconf
+    automake
+    libtool
```
Comment (markuskowa) [inline pkgs/by-name/pn/pnfft/package.nix:34]: does the `autoreconfHook` work here instead (and dropping the bootstrap script)?

File: pkgs/by-name/pf/pfft/package.nix [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  preConfigure = ''
+    export FCFLAGS="-I${lib.getDev fftw'}/include"
+    bash bootstrap.sh
+  '';
+
+  configureFlags = [
+    "--enable-portable-binary"
+  ] ++ lib.optional (precision != "double") "--enable-${precision}";
+
+  buildInputs = lib.optionals stdenv.cc.isClang [ llvmPackages.openmp ];
```
Comment (markuskowa) [inline pkgs/by-name/pf/pfft/package.nix:46]:
```suggestion
  buildInputs = lib.optional stdenv.cc.isClang llvmPackages.openmp;
```

File: pkgs/by-name/pn/pnfft/package.nix (hunk 2 above)
Comment (hmenke) [reply pkgs/by-name/pn/pnfft/package.nix:32]: There is no `NIX_FCFLAGS` that is automatically picked up and the `configure` script doesn't search for `fftw3-mpi.f03` or `pfft.f03`.

Comment (markuskowa) [reply pkgs/by-name/pn/pnfft/package.nix:32]: The compiler wrapper for gfortran should add the `-I` options automatically from the packages in `buildInputs`. I am not sure why this does not work in this case.

Comment (hmenke) [reply pkgs/by-name/pn/pnfft/package.nix:32]: Maybe because it uses `mpif90` instead of `gfortran`? I'm only guessing.

**Comment by markuskowa** [comment]
The builds fail on Darwin. Please limit the platform to Linux.

**Pull request merged**